### PR TITLE
Closes #2863 fix_join_type: Fix joined function for join_type = "before"/"after"

### LINKED ISSUE #2863: Update lintr configurations to use central `{admiraldev}` file
We are updating the way we set up linters across the `{admiral}` ecosystem.

Pull request https://github.com/pharmaverse/admiraldev/pull/507 has introduced a way to set up custom `{lintr}` configurations in `{admiraldev}` which can then used by all other packages in the `{admiral}` ecosystem. If the configurations are updated in `{admiraldev}`, they will trickle down to the `{admiral}` and the extension packages, unless they are overridden (see below). Currently the configurations we have chosen are:

- line length: 100 (same as previously)
- object usage: NULL (same as previously)
- cyclomatic complexity: 20 (same as previously)
- indentation: NULL (same as previously)
- undesirable functions: `stop()`, `warning()`, `message()`, `ifelse()`, `grepl()` 🆕

To implement the default configurations, simply make a PR for this issue where you:

- Install the development version of `{admiraldev}`
- Rename the `.lint` config file to `.lintr.R`
- Replace the contents of the file with:
```
# source in temporary environment to avoid changing the global environment
temp_env <- new.env(parent = globalenv())

source(system.file("lintr/linters.R", package = "admiraldev"), local = temp_env)

linters <- temp_env$admiral_linters(cyclocomp = cyclocomp_linter(complexity_limit = 21))

# remove temporary environment to avoid lintr warning regarding "unused settings"
rm(temp_env)

exclusions <- list(
  "R/data.R" = Inf,
  "inst" = list(undesirable_function_linter = Inf),
  "vignettes" = list(undesirable_function_linter = Inf)
)
```
- Run `lintr::lint_package()` to check the linters run as expected
- Fix any new issues with lints*; it's likely we have been using at least a few of the now undesirable functions.
- Update `.Rbuildignore` to ignore new `.lintr.R `file.

\* There may be quite a lot of changes here, depending on how well we've previously adhered to our own new guidelines. It may be worth splitting this activity across two or three programmers.

### PR DESCRIPTION
**Pull request by bundfussr**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -32,6 +36,14 @@ the derivation is called for all subsets. (#2645)
     
     **Phase 4 (removed)**
 
+- The values of the variable specified for `tmp_obs_nr_var` in
```
Comment (bms63) [inline NEWS.md:39]: should we move this up in the breaking changes, i.e. above the deprecation stuff?

File: tests/testthat/_snaps/derive_joined.md [hunk 2]
```diff
@@ -7,3 +7,20 @@
...
+    Condition
+      Warning:
+      Dataset `dataset` contains duplicate records with respect to `subj` and `day`
+      i Run `admiral::get_duplicates_dataset()` to access the duplicate records
+      Warning:
+      Dataset `dataset_add` contains duplicate records with respect to `subj` and `day`
+      i Run `admiral::get_duplicates_dataset()` to access the duplicate records
+      Warning:
+      Dataset contains duplicate records with respect to `subj`, `tmp_obs_nr_1`, and `day.join`
+      i Run `admiral::get_duplicates_dataset()` to access the duplicate records
```
Comment (bms63) [inline tests/testthat/_snaps/derive_joined.md:25]: if a user runs get_duplicates_dateset() will they see tmp_obs_nr_1 and day.join in the dataset?

File: NEWS.md (hunk 1 above)
Comment (bundfussr) [reply NEWS.md]: Yes, good idea.

File: tests/testthat/_snaps/derive_joined.md (hunk 2 above)
Comment (bundfussr) [reply tests/testthat/_snaps/derive_joined.md:25]: Yes

File: R/derive_joined.R [hunk 3]
```diff
@@ -1003,31 +1006,66 @@ get_joined_data <- function(dataset,
...
+      select(data, !!!by_vars, !!!replace_values_by_names(order)),
+      select(data_add, !!!by_vars, !!!replace_values_by_names(order))
+    ) %>%
+      distinct() %>%
       derive_var_obs_number(
         new_var = !!tmp_obs_nr_var,
         by_vars = by_vars,
         order = order,
-        check_type = check_type
+        check_type = "none"
```
Comment (bms63) [inline R/derive_joined.R:1054]: just making sure that  `check_type = "none"` is intentional?  i think i understand why but just making sure!!

Comment (bundfussr) [reply R/derive_joined.R:1054]: Yes, it is intentional. Due to the `distinct()` call before, there can't be any duplicates. Thus I want to avoid spending time on checking it.

**Review by bms63 — APPROVED** [review]
@manciniedoardo any further comments?

**Comment by manciniedoardo** [comment]
> @manciniedoardo any further comments?

in all honesty i haven't taken a look - i trust your and @bundfussr 's code/review, but if you'd like me to review additionally, i can tomorrow.
